tempesta_fw/tfw_tfwcfg.c

```diff
@@ -0,0 +1,676 @@
+/**
+ *		Tempesta FW
+ *
+ * Copyright (C) 2016 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#include "tempesta_fw.h"
+#include "tfw_tfwcfg.h"
+#include "str.h"
+
+/*
+ * Helper function that converts a constant value (enum) to a string
+ * that it corresponds to. Also there's conversion from a string to
+ * a corresponding constant. Currenty that's done via TfwCfgEnum and
+ * tfw_cfg_map_enum() function.
+ */
+static const char *
+tfw_enum_string(const char * const * enum_string, int enum_max, int enum_val)
+{
+	if ((enum_val < 0) || (enum_val >= enum_max))
+		return enum_string[enum_max];
+	if (enum_string[enum_val])
+		return enum_string[enum_val];
+	return enum_string[enum_max];
+}
+
+/* Mappings and functions for configuration directives (statements). */
+static const TfwCfgEnum const __read_mostly __tfwcfg_stmt_enum[] = {
+	{ "cache_bypass",	TFW_D_CACHE_BYPASS },
+	{ "cache_fulfill",	TFW_D_CACHE_FULFILL },
+	{}
+};
+static const char * const __read_mostly __tfwcfg_enum_stmt[] = {
+	[0 ... _TFW_D_COUNT]	= NULL,
+	[TFW_D_CACHE_BYPASS]	= "cache_bypass",
+	[TFW_D_CACHE_FULFILL]	= "cache_fulfill",
+	[_TFW_D_COUNT]		= "UNKNOWN",
+};
+const char *
+tfw_stmt_string(tfw_stmt_t stmt)
+{
+	return tfw_enum_string(__tfwcfg_enum_stmt, _TFW_D_COUNT, stmt);
```

Comment:
For now the functions are used as they used now. They complex and slow. There is no sense to introduce additional complexity and slowness w/o necessarity.